microprofile/WFLY-12590_Integrate_MicroProfile_Fault_Tolerance_2.0.adoc

```diff
@@ -0,0 +1,102 @@
+:email:             rhusar@redhat.com
+:toc:               left
+:icons:             font
+:idprefix:
+:idseparator:       -
+:keywords:          microprofile,fault-tolerance,openshift
+
+== Overview
+
+Integrate support for Eclipse MicroProfile (MP) Fault Tolerance (FT) for applications deployed in WildFly.
+This proposal pertains to MP FT version 2.0 given the timing of this RFE.
+The integration is based on SmallRye project, specifically `smallrye-fault-tolerance` module.
+The `smallrye-fault-tolerance` module is already in use by projects https://thorntail.io/[Thorntail] (WildFly Swarm) and https://quarkus.io/[Quarkus].
+
+MicroProfile Fault Tolerance specification simplifies the challenges brought by the distributed nature of microservices.
+This is because the communication with external systems is inherently unreliable.
+This increases demand on resiliency of applications.
+
+The patterns brought by this specification are annotations such as `@Timeout`, `@Retry`, `@Fallback`, `@CircuitBreaker`, `@Bulkhead` and `@Asynchronous`.
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-12590[WFLY-12590 Integrate MicroProfile Fault Tolerance 2.0]
+* https://issues.jboss.org/browse/EAP7-1347[EAP7-1347 Integrate MP Fault Tolerance 2.0 into EAP]
+
+=== Related Issues
+
+* https://issues.jboss.org/browse/WFLY-10522[WFLY-10522 Support Eclipse MicroProfile Config] - dependency; already integrated
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:mnovak@redhat.com[Miroslav Novak]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+[ ] Engineering
+
+[x] QE
+
+=== Affected Projects or Components
+
+* The specification repository resides at official Eclipse organization on GitHub in https://github.com/eclipse/microprofile-fault-tolerance[eclipse/microprofile-fault-tolerance] repository.
+* The implementation of Eclipse MicroProfile Fault Tolerance will be provided by SmallRye project https://github.com/smallrye/smallrye-fault-tolerance[smallrye/smallrye-fault-tolerance].
+* The underlying library used by _current_ SmallRye implementation is Hystrix which resides in official Netflix GitHub organization https://github.com/Netflix/Hystrix[Netflix/Hystrix].
+
+=== Other Interested Projects
+
+Not interested in our integration per se, but projects that have already integrated MP FT:
+
+* https://thorntail.io/[Thorntail] has integrated MicroProfile Fault Tolerance as a https://github.com/thorntail/thorntail/tree/master/fractions/microprofile/microprofile-fault-tolerance[fraction].
+* https://quarkus.io/[Quarkus] has integrated MicroProfile Fault Tolerance as an https://github.com/quarkusio/quarkus/tree/master/extensions/smallrye-fault-tolerance[extension].
+
+== Requirements
+
+* Support Eclipse MicroProfile Fault Tolerance version 2.0 for all applications deployed in WildFly.
```

Comment:
@rhusar This includes not only JAX-RS applications but other EE types (EJBs, Servlets,...) as well, right?

Reply:
I don't think there is any limitation where you can use this API wherever you can use CDI (Weld)

Reply:
@rhusar thanks, will be CDI required?

Reply:
@mnovak1 Most (if not all) MP specs are based on CDI.

Reply:
@rhusar the idea is to have possibility to use advantages of MP FT annotations for example in EJB (without CDI). Will it be possible? 
It's probably more question what is the trigger point for enabling MP FT in deployment?

Reply:
@mnovak1 You can do that. You just need to have the MP FT extension and empty subsystem in your profile xml. The DUPs add necessary module dependencies and CDI extension. This is in line with what the existing MP subsystems are doing.

Reply:
Just to rephrase, WildFly can't technically limit where these CDI annotations can be used since that one of the limitation in CDI. The scope of where these are supposed to be used is defined by the MP spec.

Reply:
@rhusar @kenfinnigan thanks guys for clarification. I see your point. This is important for completing test plan for this feature whether there should be mixed EE and MP test scenarios. I think this is more general question. Is there a place to bring this up and having decided for the whole MP for WF?

Reply:
@rhusar I have another question to this topic. Does it makes sense to test MP FT which package inside EAR/JAR (not just WAR)?

Reply:
> @rhusar I have another question to this topic. Does it makes sense to test MP FT which package inside EAR/JAR (not just WAR)? Is this type of deployment to be supported?
Yes.

Reply:
> @rhusar @kenfinnigan thanks guys for clarification. I see your point. This is important for completing test plan for this feature whether there should be mixed EE and MP test scenarios. I think this is more general question. Is there a place to bring this up and having decided for the whole MP for WF?

To add to the above, we give users tools but if they elect to shoot themselves in the foot, we cannot do anything about it. This is case for any API really.